trio/_path.py

```diff
@@ -1,49 +1,86 @@
+from __future__ import annotations
+
+import inspect
 import os
 import pathlib
 import sys
 import types
+from collections.abc import Awaitable, Callable, Iterable
 from functools import partial, wraps
-from typing import TYPE_CHECKING, Any
+from io import BufferedRandom, BufferedReader, BufferedWriter, FileIO, TextIOWrapper
+from typing import IO, TYPE_CHECKING, Any, BinaryIO, ClassVar, TypeVar, cast, overload
 
 import trio
+from trio._file_io import AsyncIOWrapper as _AsyncIOWrapper
 from trio._util import Final, async_wraps
 
+if TYPE_CHECKING:
+    from _typeshed import (
+        OpenBinaryMode,
+        OpenBinaryModeReading,
+        OpenBinaryModeUpdating,
+        OpenBinaryModeWriting,
+        OpenTextMode,
+    )
+    from typing_extensions import Concatenate, Literal, ParamSpec, TypeAlias
+
+    P = ParamSpec("P")
+
+T = TypeVar("T")
+StrPath: TypeAlias = "str | os.PathLike[str]"
+
 
 # re-wrap return value from methods that return new instances of pathlib.Path
-def rewrap_path(value):
+def rewrap_path(value: T) -> T | Path:
     if isinstance(value, pathlib.Path):
-        value = Path(value)
-    return value
+        return Path(value)
+    else:
+        return value
 
 
-def _forward_factory(cls, attr_name, attr):
+def _forward_factory(
+    cls: AsyncAutoWrapperType,
+    attr_name: str,
+    attr: Callable[Concatenate[pathlib.Path, P], T],
+) -> Callable[Concatenate[Path, P], T | Path]:
     @wraps(attr)
-    def wrapper(self, *args, **kwargs):
+    def wrapper(self: Path, *args: P.args, **kwargs: P.kwargs) -> Path | Any:
```

Comment:
You can use `T` instead of `Any` here. Unless there's a mypy bug (I've experienced some in this kinda place before).